### PR TITLE
refactor goto_block and also add unwind_to_block

### PR DESCRIPTION
**Pull request by RalfJung**
r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/interpret/terminator.rs [hunk 1]
```diff
@@ -272,7 +261,7 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 // No stack frame gets pushed, the main loop will just act as if the
                 // call completed.
                 if ret.is_some() {
```
Comment (oli-obk) [inline src/librustc_mir/interpret/terminator.rs]: Make this if let some and use go_to_block

File: src/librustc_mir/interpret/eval_context.rs [hunk 2]
```diff
@@ -628,10 +659,7 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         if cur_unwinding {
             // Follow the unwind edge.
             let unwind = next_block.expect("Encounted StackPopCleanup::None when unwinding!");
-            let next_frame = self.frame_mut();
```
Comment (oli-obk) [inline src/librustc_mir/interpret/eval_context.rs]: The way next_frame is used in this function makes me a bit unhappy, but I can't come up with a better scheme on the spot and it's preexisting anyway. So.. OK I guess

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: src/librustc_mir/interpret/terminator.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/terminator.rs]: This entire code here makes little sense actually... I'll adjust `call_intrinsic` to put responsibility for advancing the instruction pointer onto the hook, like we do for other `call_*` hooks.

Comment (RalfJung) [reply src/librustc_mir/interpret/terminator.rs]: I decided to change this and, uh, made the PR a lot bigger along the way... but also made the APIs a lot better, I think.

**Comment by RalfJung** [comment]
The corresponding Miri PR is at https://github.com/rust-lang/miri/pull/1077

File: src/librustc_mir/interpret/intrinsics.rs [hunk 3]
```diff
@@ -268,34 +274,39 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 // exception from the exception.)
                 // This is the dual to the special exception for offset-by-0
                 // in the inbounds pointer offset operation (see the Miri code, `src/operator.rs`).
-                if a.is_bits() && b.is_bits() {
+                //
+                // Control flow is weird because we cannot early-return (to reach the
```
Comment (oli-obk) [inline src/librustc_mir/interpret/intrinsics.rs]: we have a bunch of situations like that in miri... Maybe we should consider using `try` blocks at some point

**Pull request merged**